src/core/metadata/qgslayermetadata.cpp

```diff
@@ -0,0 +1,287 @@
+}
+
+void QgsLayerMetadata::setCrs( const QgsCoordinateReferenceSystem &crs )
+{
+  mCrs = crs;
+}
+
+QMap<QString, QStringList> QgsLayerMetadata::keywords() const
+{
+  return mKeywords;
+}
+
+void QgsLayerMetadata::setKeywords( const QMap<QString, QStringList> &keywords )
+{
+  mKeywords = keywords;
+}
+
+void QgsLayerMetadata::addKeywords( const QString &vocabulary, const QStringList &keywords )
+{
+  mKeywords.insert( vocabulary, keywords );
+}
+
+QStringList QgsLayerMetadata::keywordVocabularies() const
+{
+  return mKeywords.keys();
+}
+
+QStringList QgsLayerMetadata::keywords( const QString &vocabulary ) const
+{
+  return mKeywords.value( vocabulary );
+}
+
+QList<QgsLayerMetadata::Contact> QgsLayerMetadata::contacts() const
+{
+  return mContacts;
+}
+
+void QgsLayerMetadata::setContacts( const QList<Contact> &contacts )
+{
+  mContacts = contacts;
+}
+
+void QgsLayerMetadata::addContact( const QgsLayerMetadata::Contact &contact )
+{
+  mContacts << contact;
+}
+
+QList<QgsLayerMetadata::Link> QgsLayerMetadata::links() const
+{
+  return mLinks;
+}
+
+void QgsLayerMetadata::setLinks( const QList<QgsLayerMetadata::Link> &links )
+{
+  mLinks = links;
+}
+
+void QgsLayerMetadata::addLink( const QgsLayerMetadata::Link &link )
+{
+  mLinks << link;
+}
+
+QString QgsLayerMetadata::language() const
+{
+  return mLanguage;
+}
+
+void QgsLayerMetadata::setLanguage( const QString &language )
+{
+  mLanguage = language;
+}
+
+void QgsLayerMetadata::saveToLayer( QgsMapLayer *layer ) const
+{
+  layer->setCustomProperty( QStringLiteral( "metadata/identifier" ), mIdentifier );
+  layer->setCustomProperty( QStringLiteral( "metadata/parentIdentifier" ), mParentIdentifier );
+  layer->setCustomProperty( QStringLiteral( "metadata/language" ), mLanguage );
+  layer->setCustomProperty( QStringLiteral( "metadata/type" ), mType );
+  layer->setCustomProperty( QStringLiteral( "metadata/title" ), mTitle );
+  layer->setCustomProperty( QStringLiteral( "metadata/extent" ), QVariant::fromValue( mExtent ) );
+  layer->setCustomProperty( QStringLiteral( "metadata/abstract" ), mAbstract );
+  layer->setCustomProperty( QStringLiteral( "metadata/fees" ), mFees );
+  layer->setCustomProperty( QStringLiteral( "metadata/rights" ), mRights );
+  layer->setCustomProperty( QStringLiteral( "metadata/licenses" ), mLicenses );
+  layer->setCustomProperty( QStringLiteral( "metadata/history" ), mHistory );
+  layer->setCustomProperty( QStringLiteral( "metadata/encoding" ), mEncoding );
+  layer->setCustomProperty( QStringLiteral( "metadata/crs" ), mCrs.authid() );
+  layer->setCustomProperty( QStringLiteral( "metadata/constraints" ), QVariant::fromValue( mConstraints ) );
+  layer->setCustomProperty( QStringLiteral( "metadata/keywords" ), QVariant::fromValue( mKeywords ) );
+  layer->setCustomProperty( QStringLiteral( "metadata/contacts" ), QVariant::fromValue( mContacts ) );
+  layer->setCustomProperty( QStringLiteral( "metadata/links" ), QVariant::fromValue( mLinks ) );
```

Comment:
AFAIK, all these structs `links`,  `extent`,  `keywords` ... are not saved in the QGIS Project. Is-it expected @nyalldawson? It works for simple QString or QStringList in the XML.